From 767d817895d8da04cba526cb7ab890a3d2887f08 Mon Sep 17 00:00:00 2001
From: Mike Yuan <me@yhndnzj.com>
Date: Mon, 29 Apr 2024 15:46:32 +0800
Subject: [PATCH] pam_systemd_loadkey: add missing PAM_EXTERN


diff --git a/src/login/pam_systemd_loadkey.c b/src/login/pam_systemd_loadkey.c
index 3b4e91124a..088fd355ef 100644
--- a/src/login/pam_systemd_loadkey.c
+++ b/src/login/pam_systemd_loadkey.c
@@ -18,7 +18,7 @@
  * This can be overridden by the keyname= parameter. */
 static const char DEFAULT_KEYNAME[] = "cryptsetup";
 
-_public_ int pam_sm_authenticate(
+_public_ PAM_EXTERN int pam_sm_authenticate(
                 pam_handle_t *handle,
                 int flags,
                 int argc, const char **argv) {
@@ -89,7 +89,7 @@ _public_ int pam_sm_authenticate(
         return PAM_SUCCESS;
 }
 
-_public_ int pam_sm_setcred(
+_public_ PAM_EXTERN int pam_sm_setcred(
                 pam_handle_t *handle,
                 int flags,
                 int argc, const char **argv) {

