From 72bbd740a0df2705fd3c245c42aab702323e200e Mon Sep 17 00:00:00 2001
From: Lennart Poettering <lennart@poettering.net>
Date: Mon, 27 Nov 2023 17:10:35 +0100
Subject: [PATCH] homed: add missing bus call to homed access policy


Index: systemd/src/home/org.freedesktop.home1.conf
===================================================================
--- systemd.orig/src/home/org.freedesktop.home1.conf
+++ systemd/src/home/org.freedesktop.home1.conf
@@ -127,6 +127,10 @@
 
                 <allow send_destination="org.freedesktop.home1"
                        send_interface="org.freedesktop.home1.Manager"
+                       send_member="DeactivateAllHomes"/>
+
+                <allow send_destination="org.freedesktop.home1"
+                       send_interface="org.freedesktop.home1.Manager"
                        send_member="Rebalance"/>
 
                 <!-- Home object -->
