From 6fecdd3cc5514f2805a74bcb0de4a542c1a89966 Mon Sep 17 00:00:00 2001
From: Lennart Poettering <lennart@poettering.net>
Date: Wed, 12 Jun 2024 18:04:11 +0200
Subject: [PATCH] Merge pull request #33100 from dtardon/crypttab-parse-all

cryptsetup-generator: continue parsing after error

diff --git a/src/cryptsetup/cryptsetup-generator.c b/src/cryptsetup/cryptsetup-generator.c
index 4db25d362f..6a3ff961fc 100644
--- a/src/cryptsetup/cryptsetup-generator.c
+++ b/src/cryptsetup/cryptsetup-generator.c
@@ -795,10 +795,56 @@ static int parse_proc_cmdline_item(const char *key, const char *value, void *dat
         return 0;
 }
 
+static int add_crypttab_device(const char *name, const char *device,  const char *keyspec, const char *options) {
+        _cleanup_free_ char *keyfile = NULL, *keydev = NULL, *headerdev = NULL, *filtered_header = NULL;
+        crypto_device *d = NULL;
+        char *uuid;
+        int r;
+
+        uuid = startswith(device, "UUID=");
+        if (!uuid)
+                uuid = path_startswith(device, "/dev/disk/by-uuid/");
+        if (!uuid)
+                uuid = startswith(name, "luks-");
+        if (uuid)
+                d = hashmap_get(arg_disks, uuid);
+
+        if (arg_allow_list && !d) {
+                log_info("Not creating device '%s' because it was not specified on the kernel command line.", name);
+                return 0;
+        }
+
+        r = split_locationspec(keyspec, &keyfile, &keydev);
+        if (r < 0)
+                return r;
+
+        if (options && (!d || !d->options)) {
+                r = filter_header_device(options, &headerdev, &filtered_header);
+                if (r < 0)
+                        return r;
+                options = filtered_header;
+        }
+
+        r = create_disk(name,
+                        device,
+                        keyfile,
+                        keydev,
+                        (d && d->options) ? d->headerdev : headerdev,
+                        (d && d->options) ? d->options : options,
+                        arg_crypttab);
+        if (r < 0)
+                return r;
+
+        if (d)
+                d->create = false;
+
+        return 0;
+}
+
 static int add_crypttab_devices(void) {
         _cleanup_fclose_ FILE *f = NULL;
         unsigned crypttab_line = 0;
-        int r;
+        int r, ret = 0;
 
         if (!arg_read_crypttab)
                 return 0;
@@ -811,10 +857,7 @@ static int add_crypttab_devices(void) {
         }
 
         for (;;) {
-                _cleanup_free_ char *line = NULL, *name = NULL, *device = NULL, *keyspec = NULL, *options = NULL,
-                                    *keyfile = NULL, *keydev = NULL, *headerdev = NULL, *filtered_header = NULL;
-                crypto_device *d = NULL;
-                char *uuid;
+                _cleanup_free_ char *line = NULL, *name = NULL, *device = NULL, *keyspec = NULL, *options = NULL;
                 int k;
 
                 r = read_stripped_line(f, LONG_LINE_MAX, &line);
@@ -834,49 +877,14 @@ static int add_crypttab_devices(void) {
                         continue;
                 }
 
-                uuid = startswith(device, "UUID=");
-                if (!uuid)
-                        uuid = path_startswith(device, "/dev/disk/by-uuid/");
-                if (!uuid)
-                        uuid = startswith(name, "luks-");
-                if (uuid)
-                        d = hashmap_get(arg_disks, uuid);
-
-                if (arg_allow_list && !d) {
-                        log_info("Not creating device '%s' because it was not specified on the kernel command line.", name);
-                        continue;
-                }
-
-                r = split_locationspec(keyspec, &keyfile, &keydev);
-                if (r < 0)
-                        return r;
-
-                if (options && (!d || !d->options)) {
-                        r = filter_header_device(options, &headerdev, &filtered_header);
-                        if (r < 0)
-                                return r;
-                        free_and_replace(options, filtered_header);
-                }
-
-                r = create_disk(name,
-                                device,
-                                keyfile,
-                                keydev,
-                                (d && d->options) ? d->headerdev : headerdev,
-                                (d && d->options) ? d->options : options,
-                                arg_crypttab);
-                if (r < 0)
-                        return r;
-
-                if (d)
-                        d->create = false;
+                RET_GATHER(ret, add_crypttab_device(name, device, keyspec, options));
         }
 
-        return 0;
+        return ret;
 }
 
 static int add_proc_cmdline_devices(void) {
-        int r;
+        int r, ret = 0;
         crypto_device *d;
 
         HASHMAP_FOREACH(d, arg_disks) {
@@ -902,11 +910,10 @@ static int add_proc_cmdline_devices(void) {
                                 d->headerdev,
                                 d->options ?: arg_default_options,
                                 "/proc/cmdline");
-                if (r < 0)
-                        return r;
+                RET_GATHER(ret, r);
         }
 
-        return 0;
+        return ret;
 }
 
 DEFINE_PRIVATE_HASH_OPS_WITH_VALUE_DESTRUCTOR(crypt_device_hash_ops, char, string_hash_func, string_compare_func,
@@ -932,14 +939,9 @@ static int run(const char *dest, const char *dest_early, const char *dest_late)
                 return 0;
 
         r = add_crypttab_devices();
-        if (r < 0)
-                return r;
-
-        r = add_proc_cmdline_devices();
-        if (r < 0)
-                return r;
+        RET_GATHER(r, add_proc_cmdline_devices());
 
-        return 0;
+        return r;
 }
 
 DEFINE_MAIN_GENERATOR_FUNCTION(run);

