From 75fb956f70513669a3f09bab93c3eb76d72dbe4f Mon Sep 17 00:00:00 2001
From: Lennart Poettering <lennart@poettering.net>
Date: Wed, 8 May 2024 13:08:38 +0200
Subject: [PATCH] bus-util: check sd_bus_pending_method_calls() when
 determining whether a service is idle


diff --git a/src/shared/bus-util.c b/src/shared/bus-util.c
index 30f9602b1e..216c16f039 100644
--- a/src/shared/bus-util.c
+++ b/src/shared/bus-util.c
@@ -122,7 +122,9 @@ int bus_event_loop_with_idle(
                 if (r == SD_EVENT_FINISHED)
                         break;
 
-                if (check_idle)
+                if (ordered_hashmap_size(bus->reply_callbacks) > 0)
+                        idle = false;
+                else if (check_idle)
                         idle = check_idle(userdata);
                 else
                         idle = true;

