From 16bae207c2b2efef82799ded8732a955555a5191 Mon Sep 17 00:00:00 2001
From: Mike Yuan <me@yhndnzj.com>
Date: Mon, 15 Dec 2025 21:02:15 +0100
Subject: [PATCH] core/dbus-execute: respect write flags for empty assignments
 too


diff --git a/src/core/dbus-execute.c b/src/core/dbus-execute.c
index b7d8b4b5ff..bf36a928b1 100644
--- a/src/core/dbus-execute.c
+++ b/src/core/dbus-execute.c
@@ -3914,7 +3914,7 @@ int bus_exec_context_set_transient_property(
                 if (r < 0)
                         return r;
 
-                if (empty) {
+                if (!UNIT_WRITE_FLAGS_NOOP(flags) && empty) {
                         bind_mount_free_many(c->bind_mounts, c->n_bind_mounts);
                         c->bind_mounts = NULL;
                         c->n_bind_mounts = 0;
@@ -3959,7 +3959,7 @@ int bus_exec_context_set_transient_property(
                 if (r < 0)
                         return r;
 
-                if (empty) {
+                if (!UNIT_WRITE_FLAGS_NOOP(flags) && empty) {
                         temporary_filesystem_free_many(c->temporary_filesystems, c->n_temporary_filesystems);
                         c->temporary_filesystems = NULL;
                         c->n_temporary_filesystems = 0;

