From c98c0ba4c5262d101475a3f9d05bb1b649c79708 Mon Sep 17 00:00:00 2001
From: Yu Watanabe <watanabe.yu+github@gmail.com>
Date: Thu, 28 Mar 2024 13:13:20 +0900
Subject: [PATCH] network/ndisc: NDisc Route option with 0 prefixlength is
 valid

Fixes #31606.

diff --git a/src/network/networkd-ndisc.c b/src/network/networkd-ndisc.c
index 61d591a2af..264ecff34d 100644
--- a/src/network/networkd-ndisc.c
+++ b/src/network/networkd-ndisc.c
@@ -877,11 +877,6 @@ static int ndisc_router_process_route(Link *link, sd_ndisc_router *rt) {
         if (r < 0)
                 return log_link_warning_errno(link, r, "Failed to get route prefix length: %m");
 
-        if (in6_addr_is_null(&dst) && prefixlen == 0) {
-                log_link_debug(link, "Route prefix is ::/0, ignoring");
-                return 0;
-        }
-
         if (in6_prefix_is_filtered(&dst, prefixlen,
                                    link->network->ndisc_allow_listed_route_prefix,
                                    link->network->ndisc_deny_listed_route_prefix)) {

