From 87333bd1dc69195b93e9aee9b91c06fb167b152e Mon Sep 17 00:00:00 2001
From: Daan De Meyer <daan.j.demeyer@gmail.com>
Date: Fri, 4 Oct 2024 21:34:33 +0200
Subject: [PATCH] chase: Fix shortcut

We can't shortcut chaseat() if CHASE_PARENT is set.

diff --git a/src/basic/chase.c b/src/basic/chase.c
index 4576e4b058..8eac356665 100644
--- a/src/basic/chase.c
+++ b/src/basic/chase.c
@@ -192,7 +192,7 @@ int chaseat(int dir_fd, const char *path, ChaseFlags flags, char **ret_path, int
 
         if (!(flags &
               (CHASE_AT_RESOLVE_IN_ROOT|CHASE_NONEXISTENT|CHASE_NO_AUTOFS|CHASE_SAFE|CHASE_STEP|
-               CHASE_PROHIBIT_SYMLINKS|CHASE_MKDIR_0755)) &&
+               CHASE_PROHIBIT_SYMLINKS|CHASE_MKDIR_0755|CHASE_PARENT)) &&
             !ret_path && ret_fd) {
 
                 /* Shortcut the ret_fd case if the caller isn't interested in the actual path and has no root

