From 3b5495c0416c21679f0cca1fbaabb22e093c6c58 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Zbigniew=20J=C4=99drzejewski-Szmek?= <zbyszek@amutable.com>
Date: Fri, 15 May 2026 18:45:01 +0200
Subject: [PATCH] homectl: fix unlocking of multiple homes

Fixes 4aa0a8ac3e548219d380a0c566242338eab185da.

diff --git a/src/home/homectl.c b/src/home/homectl.c
index 42b6359fc6..75339d93c8 100644
--- a/src/home/homectl.c
+++ b/src/home/homectl.c
@@ -2788,7 +2788,7 @@ static int verb_unlock_home(int argc, char *argv[], uintptr_t _data, void *userd
 
                         r = sd_bus_call(bus, m, HOME_SLOW_BUS_CALL_TIMEOUT_USEC, &error, NULL);
                         if (r < 0) {
-                                r = handle_generic_user_record_error(argv[1], secret, &error, r, false);
+                                r = handle_generic_user_record_error(*i, secret, &error, r, false);
                                 if (r < 0) {
                                         if (ret == 0)
                                                 ret = r;

